core/src/main/java/de/danoeh/antennapod/core/storage/ItemEnqueuePositionCalculator.java

```diff
@@ -0,0 +1,106 @@
+ */
+class ItemEnqueuePositionCalculator {
+
+    public static class Options {
+        private boolean enqueueAtFront = false;
+        private boolean keepInProgressAtFront = false;
+
+        public boolean isEnqueueAtFront() {
+            return enqueueAtFront;
+        }
+
+        public Options setEnqueueAtFront(boolean enqueueAtFront) {
+            this.enqueueAtFront = enqueueAtFront;
+            return this;
+        }
+
+        public boolean isKeepInProgressAtFront() {
+            return keepInProgressAtFront;
+        }
+
+        public Options setKeepInProgressAtFront(boolean keepInProgressAtFront) {
+            this.keepInProgressAtFront = keepInProgressAtFront;
+            return this;
+        }
+    }
+
+    @NonNull
+    private final Options options;
+
+    @VisibleForTesting
+    DownloadStateProvider downloadStateProvider = DownloadRequester.getInstance();
+
+    public ItemEnqueuePositionCalculator(@NonNull Options options) {
+        this.options = options;
+    }
+
+    /**
+     *
+     * @param positionAmongToAdd Typically, the callers has a list of items to be inserted to
+     *                           the queue. This parameter indicates the position (0-based) of
+     *                           the item among the one to inserted. E.g., it is needed for
+     *                           enqueue at front option.
+     *
+     * @param item the item to be inserted
+     * @param curQueue the queue to which the item is to be inserted
+     * @return the position (0-based) the item should be inserted to the named queu
+     */
+    public int calcPosition(int positionAmongToAdd, FeedItem item, List<FeedItem> curQueue) {
+        if (options.isEnqueueAtFront()) {
+            if (options.isKeepInProgressAtFront() &&
+                    curQueue.size() > 0 &&
+                    curQueue.get(0).getMedia() != null &&
+                    curQueue.get(0).getMedia().isInProgress()) {
+                // leave the front in progress item at the front
+                return getPositionOf1stNonDownloadingItem(positionAmongToAdd + 1, curQueue);
+            } else { // typical case
+                // return NOT 0, so that when a list of items are inserted, the items inserted
+                // keep the same order. Returning 0 will reverse the order
+                return getPositionOf1stNonDownloadingItem(positionAmongToAdd, curQueue);
+            }
+        } else {
+            return curQueue.size();
+        }
+    }
+
+    private int getPositionOf1stNonDownloadingItem(int startPosition, List<FeedItem> curQueue) {
+        final int curQueueSize = curQueue.size();
+        for (int i = startPosition; i < curQueueSize; i++) {
+            if (!isItemAtPositionDownloading(i, curQueue)) {
+                return i;
+            } // else continue to search;
+        }
+        return curQueueSize;
+    }
+
+    private boolean isItemAtPositionDownloading(int position, List<FeedItem> curQueue) {
```

Comment:
Couldn't the queue be a field of the class instead of passing it to every method?

Reply:
Right now, the main entry point method the the class is:
```java
  int calcPosition(positionAmongToAdd, feedItem, curQueue)
```
The method signature suggests the function calculates the position of the feedItem in the given queue.

Moving the queue to be a member of the class would make the main entry point method to 
```java
  int calcPosition(positionAmongToAdd, feedItem)
```
It is less readable.

Reply:
This whole class is kind of complicated. A lot of looping through the queue, for each item. I have another suggestion that might be easier to read and understand that could be added to `DBWriter.java`:

```java
int indexToAddEpisode;
switch (enqueueLocation) {
    case BACK:
        indexToAddEpisode = queue.size();
        break;
    case FRONT:
        indexToAddEpisode = 0;
        break;
    case AFTER_CURRENTLY_PLAYING:
        indexToAddEpisode = getCurrentlyPlaying(queue);
        break;
}

for (int i = 0; i < itemIds.length; i++) {
    // old code in DBWriter.java
    queue.add(indexToAddEpisode, itemId);
    indexToAddEpisode++;
}
```

Reply:
The basic idea would work. I'll get back to you.

Reply:
If I understood the behavior correctly, the position will start at some point and will then just be incremented for each new item. Your version re-calculates the position again and again (please note the `indexToAddEpisode++;` in my example). I think that's bad for performance (does not really matter) but it's also hard to understand. When just incrementing the index, you directly understand that the items are added in bulk.

Reply:
Please review the latest commit. If we agree on the direction,
- I'll add some tests to cover the implementation.
- What about replacing this PR with another one? After the reviews, both backend and UI change considerably, making the version history somewhat complicated. 
I can start another branch / PR that includes the final option.

Reply:
> I can start another branch / PR that includes the final option.

Would be great to have some clean-up before merging. Force pushing a cleaned up/squashed history would be another option.

Reply:
I'm afraid the most recent commit (using `insertIndex++` to support download order) won't quite work for #2448. I remember the original reason for the logic of looking for downloading episodes.

#2448, respect download start order, should work  when user _manually_ taps download _in succession_.  Example:
- assume enqueue location is at front,
- the current queue is `A, B, C`
- the user wants to download episode D and episode E.
- user taps to download episode D
- queue should be: `D, A, B, C`
- user taps to download episode E (D is still downloading)
- queue should be: `D, E, A, B, C`

Given the two download taps map to two separate `addQueueItem()` calls,
- with the `insertIndex++` approach, the resulting queue would be `E, D, A, B, C`, the opposite of the user's order.
- the logic of looking over downloading episodes are to handle such cases, so that when user taps to download episode E, it will try to add to the front, but after `D`, as `D` is being downloaded.|

Reply:
Can't this also be fixed by returning a specific starting position in `calcInsertIndex`?

Reply:
It could, but the looping for downloading episodes logic would carry over, roughly equal to calling the calculator class once outside the loop.

Reply:
I still think this is easier to understand than calling the calculator class in every loop. You can directly see that it just cares about the starting position. Just a personal feeling.

Reply:
I get what you mean, and will adjust it accordingly.

I recalled a (historical) reason why the calculator class was called inside the loop. The intent (back then) was to possibly support feeds priority: individual feeditem would be enqueued at different position. Since feed priority is not on radar screen, we don't really need it at this point.

Reply:
Please review the latest commit. It reflects the discussion.

If there is no more major changes, I'd like to do 2 things before saying the PR is ready for merge:
1. minor code style fixes on existing code in the section I've touched, e.g., changing the case parameter in `DBWriterTest. queueTestSetupMultipleItems(final int NUM_ITEMS) `
2. simplify version history with squash / reorder.

Reply:
@ByteHamster is there anything you'd like to raise regarding the latest commit, before I make the code style changes / version history squash + reorder?

(I didn't realize you've put a thumbs-up on my previous comment, and can't tell it it means "ok, go ahead and make the changes," or just a general reaction to the comment.)